### PR TITLE
[training_utils] feat: impl a metrics class for recording metrics and support worker sync [WIP]

### PR DESCRIPTION
**Pull request by 0x404**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### High-Level Design

> Demonstrate the high-level design if this PR is complex.

### Specific Changes

> List the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ).

### CODE REVIEW

**Comment by 0x404** [comment]
Hi @vermouth1992, I found the approach we discussed in https://github.com/volcengine/verl/pull/2259#discussion_r2173753481 is somewhat tricky. We need more discussion to support `aggregated_data` for different dispatch_fn and collect_fn. And since we are going to deprecate `DataProto` in the future refactor, I think we shouldn't support `aggregated_data` for it anymore.

I think we could support a generic `Metrics` class, and we can set a config for each recorded metric, including the reduce type it should use, and currently also including the format it should use (this is to support console log, since currently each metric can only support the default `"{:.3f}"`, so learning rate always shows as 0.000). We can support synchronization and reduce between different ranks for this Metrics class, and can use this util to replace the currently used dict for recording metrics.

This PR is still work in progress. Do you think this approach is reasonable? It would be great to hear your insights.

File: verl/utils/metric/metric.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
+from collections import defaultdict
```
Comment (vermouth1992) [inline verl/utils/metric/metric.py:1]: How to we want to aggregate among dp?

Comment (vermouth1992) [reply verl/utils/metric/metric.py:1]: I guess we need a function to merge them? Actually, I guess allgather is a good choice because this is what typically done in pretraining.

Comment (0x404) [reply verl/utils/metric/metric.py:1]: Yes, I was thinking of creating a function `allgather_across_group(self, group)` for Metrics class, which would gather all other ranks' metrics objects on each dp rank, then merge them into a single Metrics through `Metrics.merge`.

What do you think? This PR is still a draft demo, if this makes sense, I will keep working on this.
